content/languages/c/authoring/memory-management-techniques.md

```diff
@@ -6,19 +6,341 @@ sidebar: "language:c"
+                //... protect agains ill-behaving solutions which are not able to get the correct status
+            }
+        }
+        
+        //You can perform assertions on the status and returned size here, or
+        //you can create a separate suite(s) just to test them separately
+
+        //...perform assertions, verify correctness of returned numbers...
+    }
+
+    //release the memory after the test
+    free(array);
+}
+```
+
+
+### Memory managed by the solution
+
+Opposite of memory managed by the test suite is the approach of pushing the responsibility to the user. This way, tests do not need to worry about problematic aspects of memory management, kata authors give freedom of implementation to users, and can reduce the boilerplate required to implement memory management.
+
+
+#### Symmetric functions for allocation and deallocation
+
+This idea basically boils down to asking users to provide their equivalents of allocation and deallocation functions. Solution function is responsible not only for solving the task, but also for allocation of the memory, and storing of book-keeping information. Clean-up function is responsible for releasing resources.
+
+There's many possible ways of implementing this approach, and it usually ends up being similar to the [naive approach](#naive-approach-malloc-in-the-solution-and-free-in-tests) described in the beginning. As it is very useful for more complex memory structures, a couple of examples can be found in the section on [Two-dimensional arrays](#two-dimensional-arrays).
+
+
 ## Two-dimensional arrays
 
-- N+1 allocations
-- Flat
+Some kata require the user solution to return a two-dimensional array, for example a 2D matrix, or an array of C-strings. Such scenarios are a bit more complex, because not only the higher order array has to be properly managed, but all its individual entries as well. Exact approach selected for allocation of such structure depends on the scenario, because different techniques are suitable for square or rectangular arrays, jagged arrays, arrays of null-terminated strings, etc.
+
+### Naive approach: N+1 allocations
+
+
+
+### Flat array
+
+Very often overlooked, but a very good approach to represent 2D arrays is to store them in a regular, linear array of `T[ ]`. It can be effectively used when bounds between inner arrays can be efficiently determined, for example each row of a matrix has a well known length, rows of a Pascal's triangle have precisely defined, although different, lengths, and string entries are clearly terminated.
+
+This way, complexity of memory management is greatly reduced, because whole necessary memory can be allocated and freed with a single call to `malloc` (or equivalent) and `free`.
+
+Drawbacks of this approach is that the interface of the solution does not resemble its logical structure, i.e. elements of a matrix cannot be accessed with, for example, `matrix[row][col]`, but with `matrix[row * size + col]`.
```

Comment:
`matrix[row][col]` can be used, it just require a type cast. C99 is more than 20 years old, so it's not even _modern_ C, it's just C.

Reply:
Oh, this is interesting. Could you show me then how it would work in such case then?

```c
int w = rand() % 100 + 10, h = rand() % 50 + 10;
char* maze_buffer = malloc(w * h + 1);
generate_maze(maze_buffer, w, h); //invoke user solution

//how can I cast the linear maze_buffer to 2D maze?
??? maze = (???)maze_buffer;  //how to cast it?
char cell = maze[5][3];

//....
free(maze_buffer);
```

I stopped working with C before 2005, so even C99 was not widespread across compilers back then :D Thanks!

Reply:
`char (*maze)[w] = (char(*)[w])maze_buffer;` for an existing buffer, just `char maze[h][w];` for allocation on stack, `void generate_maze(size_t w, size_t h, char maze_buffer[h][w])` as a function argument (dimensions have to be before the array), asterisks may be used in prototypes like `void generate_maze(size_t, size_t, char maze_buffer[*][*]` when there are no named arguments.

Reply:
Oh great, thank you. I was not aware that it's possible to use variables in type specifiers. Smarter every day!

Reply:
I cannot get to declaring the return type of the function. I think it has to stay as `char*` and casts have to be done by the party accessing it, i.e. by solution to fill in, and by the test suite to read it?

Reply:
Actually, both dimensions can be in the type, just as with regular arrays: `char (*maze)[h][w] = (char(*)[h][w])maze_buffer;`. As for returning it, I don't think it's possible, it still is a low level language.

Reply:
I added a note on casting between linear and 2d array.